File: bench/plots/cdf.py
import sys
import pandas as pd
import matplotlib.pyplot as plt

if len(sys.argv) < 2:
    print(f"usage: {sys.argv[0]} /path/to/file.csv")
    exit(1)

csv_file = sys.argv[1]

df = pd.read_csv(csv_file)
df = df.sort_values(by="rtime")
n_rows = len(df)
df = df[df['rtime'] < 300]
df['cumulative_markers'] = [ (i / n_rows) * 100 for i in range(1, len(df) + 1)]
plt.figure(figsize=(14, 8))
plt.step(df['rtime'], df['cumulative_markers'], where='post', label='FAST', linewidth=5.0)
# Set the x and y axis to start at zero and remove margins
plt.xlim(left=0)  # X-axis starts at 0
plt.ylim(bottom=0)  # Y-axis starts at 0
#plt.margins(x=0, y=0)  # Remove any margins around the plot
plt.xlabel('Time (s)', fontsize=36)
plt.ylabel('Percentage of finished files [%]', fontsize=30)
plt.minorticks_on()
plt.grid(True, which='major', linestyle='-', linewidth=2.0)
# plt.grid(True, which='minor', linestyle=':', linewidth=2.0)

# Remove the top and right frame (spines)
ax = plt.gca()  # Get current axes
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['bottom'].set_linewidth(2.0)
ax.spines['left'].set_linewidth(2.0)
ax.tick_params(axis='x', width=2.0)
ax.tick_params(axis='y', width=2.0)
ax.tick_params(axis='x', which='minor', width=2)
ax.tick_params(axis='y', which='minor', width=2)

plt.xticks(fontsize=30)
plt.yticks(fontsize=30)
plt.legend(loc='lower right', fontsize=36)

plt.tight_layout()
plt.savefig("cdf.pdf")

File: bench/plots/cdf.ml
open Owl
open Owl_plplot

(* Meh *)
let () =
  let df = Dataframe.of_csv ~sep:',' "fast-vulcan-secbench-results.csv" in
  let df = Dataframe.sort df "rtime" in
  let n = Dataframe.row_num df in

  let rtime = Dataframe.(unpack_float_series @@ get_col_by_name df "rtime") in

  let x = Mat.init n 1 (fun i -> rtime.(i)) in
  let y = Mat.init n 1 (fun i -> (float i +. 1.) /. float n *. 100.) in

  let h = Plot.create "cdf.pdf" in

  Plot.stairs ~spec:[ RGB (70, 130, 180); LineWidth 10.; LineStyle 2 ] ~h x y;

  Plot.set_font_size h 24.;
  Plot.set_pen_size h 8.;
  Plot.set_xlabel h "Time (s)";
  Plot.set_ylabel h "Percentage of finished files [%]";


  Plot.legend_on h ~position:SouthEast [| "FAST" |];
  Plot.output h
